lib/AST/Decl.cpp

```diff
@@ -5518,6 +5519,16 @@ VarDecl::VarDecl(DeclKind kind, bool isStatic, VarDecl::Introducer introducer,
   Bits.VarDecl.IsTopLevelGlobal = false;
 }
 
+int VarDecl::getFieldIndex() const {
+  if (fieldIndex < 0) {
+    NominalTypeDecl *parent = getDeclContext()->getSelfNominalTypeDecl();
+    evaluateOrDefault(getASTContext().evaluator,
+                      ComputeFieldIndicesRequest{parent}, {});
+    assert(fieldIndex >= 0);
+  }
```

Comment:
We generally want to try and avoid having side-effecting requests like this. Ideally `ComputeFieldIndicesRequest` would be a `SeparatelyCached` cached request, and get and set the cached value in its `getCachedResult`/`cacheResult` methods.

You could also fill in the other field indices for the other properties as you iterate by calling `cacheOutput` on the evaluator for those properties, similar to what `DefaultArgumentInitContextRequest` does.

Reply:
Hm, would a better approach here be to remove `field` from the decl and just store everything in a cached `DenseMap` instead? That would use more memory but at least wouldn't have any side-effects.

Reply:
If it's going into a DenseMap then that would basically be more or less equivalent to just using the evaluator's internal cache, which we could possibly do here.

Reply:
Well, I could also add a DenseMap for this in SILModule without using the evaluator machinery. But the field indices are something which are used in a critical path in the optimizer (for computing projections). So the purpose of this exercise is to avoid the overhead of a hash lookup.

Reply:
Ideally, the index of all stored properties should be computed and cached for a type during a single request. This *should* result in a unique index for each property in an object. This is something that we can't currently rely on but would greatly simplify analysis because the index would have meaning independent of the address type.
@eeckstein is right that the index should be directly loadable from an instruction without any hash lookup. That was the point of the instruction cache, although I'm happy to see it go.